Fine tune sanity

bcoca opened the pull request:
##### SUMMARY

 - allow user finer grained control on behaviour
 - adjust regex to use native chargroups so it is more py2/3 compatible

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
inventory

bcoca committed:
moar options fer users

bcoca committed:
modify regex to use implicit charsets

  this should solve issues in py3 and unicode names

s-hertel commented on lib/ansible/inventory/group.py:
```diff
@@ -35,12 +35,13 @@ def to_safe_group_name(name, replacer="_", force=False, silent=False):
+                if not (silent or C.TRANSFORM_INVALID_GROUP_CHARS == 'silently'):
                     display.warning('Replacing ' + msg)
-            else:
+            elif C.TRANSFORM_INVALID_GROUP_CHARS != 'ignore':
```
`s/ignore/never`?

s-hertel replied:
Oh, maybe you intended for a 4th option, in which case should be added to choices.

s-hertel replied:
It seems like an 'ignore' option will cause issues with deprecation.

bcoca replied:
left it in code, JIC, but didn't want to allow it in options 'yet'

bcoca committed:
fix default

bcoca committed:
adjusted desc

s-hertel commented on lib/ansible/constants.py:
```diff
@@ -119,7 +119,8 @@ def __getitem__(self, y):
 
 # This matches a string that cannot be used as a valid python variable name i.e 'not-valid', 'not!valid@either' '1_nor_This'
-INVALID_VARIABLE_NAMES = re.compile(r'^[^a-zA-Z_]|[^a-zA-Z0-9_]')
+INVALID_VARIABLE_NAMES = re.compile(r'^[\d\W]|[^\w]')
```
Looks good. This fixes identifying valid python-3 unicode variables as invalid characters for me. cc @kdelee since I think you were testing this.

bcoca replied:
it was @felixfontein that brought it up iirc, in irc

s-hertel replied:
Oh. @kdelee brought it up in slack too

felixfontein replied:
I think it wasn't me who brought it up, I didn't mention it (I think) and I saw it in some other issue/PR.

bcoca replied:
resmo then? i created test from irc conversation this morning, that is what i've been using as lithmus

bcoca committed:
fix issue with subgroups in yaml inventory

kdelee commented:
I tried this out. I think it is an improvement and more in the spirit of "legal variable names" and allowing people to use their whole legal character set.

kdelee commented:
```
ansible-config dump --only-changed
INVENTORY_ENABLED(/home/elijah/tower/3_5_notes/inventory_plugins/ansible.cfg) = ['ini']
```
```
cat my_ini_inv.ini 
[hello]
example.com

[group-똼팇쇴]
boom.com
```


I am wondering if the default got swapped though. On the same inventory, with devel I get:
```
[DEPRECATION WARNING]: Ignoring invalid character(s) "{'똼', '팇', '쇴', '-'}" in group name (group-똼팇쇴). This feature will be removed in version 2.12. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
```
on this branch I get:
```
 [WARNING]: Replacing invalid character(s) "{'-'}" in group name (group-똼팇쇴)
```

felixfontein reviewed (commented):
Hmm, from how I understand this PR, it changes the current behavior from "warn, but don't change" to "warn, but change", and the user must manually change it to `ignore` to get "don't warn and don't change". This will break a lot of playbooks if people don't set this option explicitly to `ignore`, and if they set it to `ignore`, they don't easily find out where they still have invalid group names.

How about adding a fourth option between `always` and `ignore`, like `warn`, which warns but does not change, and make that one the default? Also, I think the warning message should (at least for the default setting) mention this option, so people know that they can change the behavior.

Edit: now I'm really confused, I see that there are choices `always`, `warn` and `never`, but the code mentions `ignore`. I could have sworn that I read the list as `always`, `ignore` and `never`. Anyway, `never` seems to do what I wanted with `warn`, so if `ignore` is allowed too and the default is `never`, I'm happy.

felixfontein commented:
@kdelee the default definitely changed. It does warn *and* modify, whereas the old default warned and did *not* modify.

s-hertel commented on lib/ansible/config/base.yml:
```diff
@@ -1441,15 +1441,17 @@ INTERPRETER_PYTHON_FALLBACK:
 TRANSFORM_INVALID_GROUP_CHARS:
   name: Transform invalid characters in group names
-  default: False
+  default: 'always'
```
I think this needs to be 'never' for backwards compatibility. That fixes https://github.com/ansible/ansible/pull/53544#issuecomment-471074131.

bcoca committed:
never say never

felixfontein commented on lib/ansible/config/base.yml:
```diff
@@ -1441,15 +1441,17 @@ INTERPRETER_PYTHON_FALLBACK:
-    - When 'true' it will replace any invalid charachters with '_' (underscore).
+    - If 'never' it will allow for the group name but warn about the issue.
+    - When 'always' it will replace any invalid charachters with '_' (underscore) and warn the user
+    - When 'silently', it does the same as 'always' sans the warnings.
```
Please add `ignore` as well!

bcoca replied:
still not convinced that should be an option

felixfontein replied:
So you allow us to keep the dashes, but only if we accept that every time we run a playbook, we get a ton of deprecation warnings with no way to turn them off (except by turning *all* deprecation warnings off)? I'm currently getting a screen full of deprecation warnings at the beginning of every playbook run with `devel`. That's really annoying.

bcoca replied:
i need to separate the deprecation warning with the per instance warning, deprecation will only exist for 2 versions till we flip the default

felixfontein replied:
Except for people who explicitly set the setting to the current default. Anyway, would it be possible to at least reduce the amounts of warnings, by printing one warning in total (or one per inventory plugin)? I guess it should contain a list of all invalid group names, which can be a very long warning if you have a lot of such groups, but at least it will take a lot of less screen space than the current huge list of warnings.

bcoca replied:
try now

bcoca committed:
clarify deprecation

bcoca committed:
separated per name warning from deprecation

bcoca committed:
bestest practices

AlanCoding commented:
I would like to be notified of changes like this. I hope it's obvious that I am already consuming this setting, and the way I do that will have to change in response to this PR.

Beyond the obvious, there are also extremely subtle changes that affect related work like https://github.com/ansible/ansible/pull/53458. With 53458 alone, the global sanitization method behavior is the same as the script with `replace_dash_in_groups=True`. With 53458+53544, the default behavior allows unicode, creating an entirely new divergence from the script behavior. That might not sound like a big deal to you, but I already have code to consume the new setting based on assumptions you're changing here.

Every divergence from behavior in the script is very important to us.

It sounds like a good change to allow unicode, for your users, but that decision is entirely in your court.

Could somebody clarify to me what the new _default_ behavior will be after this change?

bcoca committed:
finsdf

s-hertel commented:
@AlanCoding FYI, I messaged you directly about this PR before I reviewed it since it was the solution to your comment: https://github.com/ansible/ansible/pull/52748#issuecomment-470601091.

AlanCoding commented:
Yes, you have been amazingly helpful, and I greatly appreciate it. I wanted to get github notifications too. That came across much more antagonistic than I intended it to, sorry about that.

bcoca committed:
move noise to verbosity, simplify warnings

bcoca committed:
dedot

bcoca committed:
push details to v^4

 cause some people seem to always use -v3 and now don't want extra data .. that they are specifically asking for ...
